drivers/rtc/rtc_rv3028.c

```diff
@@ -0,0 +1,1101 @@
+	                        clkout);
+	if (err) {
+		return err;
+	}
+
+	/* Use edge interrupts to avoid multiple GPIO IRQs while servicing the IRQ in the thread */
+	err = gpio_pin_interrupt_configure_dt(&config->gpio_int,
+	                                      enable ? GPIO_INT_EDGE_TO_ACTIVE : GPIO_INT_DISABLE);
+	if (err) {
+		LOG_ERR("Failed to %s GPIO IRQ (err %d)", enable ? "enable" : "disable", err);
+		return err;
+	}
+
+	return 0;
+}
+
+static void rv3028_isr_thread(const struct device *dev)
+{
+	struct rv3028_data *data = dev->data;
+	rtc_alarm_callback alarm_callback = NULL;
+	void *alarm_user_data = NULL;
+	rtc_update_callback update_callback = NULL;
+	void *update_user_data = NULL;
+	uint8_t status;
+	int err;
+
+	for (;;) {
+		k_sem_take(&data->int_sem, K_FOREVER);
+		k_mutex_lock(&data->lock, K_FOREVER);
+
+		err = rv3028_read_reg8(dev, RV3028_REG_STATUS, &status);
+		if (err) {
+			goto unlock;
+		}
+
+#ifdef CONFIG_RTC_ALARM
+		if ((status & RV3028_STATUS_AF) != 0 && data->alarm_callback != NULL) {
+			status &= ~(RV3028_STATUS_AF);
+			alarm_callback = data->alarm_callback;
+			alarm_user_data = data->alarm_user_data;
+		}
+#endif /* CONFIG_RTC_ALARM */
+
+#ifdef CONFIG_RTC_UPDATE
+		if ((status & RV3028_STATUS_UF) != 0 && data->update_callback != NULL) {
+			status &= ~(RV3028_STATUS_UF);
+			update_callback = data->update_callback;
+			update_user_data = data->update_user_data;
+		}
+#endif /* CONFIG_RTC_UPDATE */
+
+		err = rv3028_write_reg8(dev, RV3028_REG_STATUS, status);
+		if (err) {
+			goto unlock;
+		}
+
+		/* Check if interrupt occurred between STATUS read/write */
+		err = rv3028_read_reg8(dev, RV3028_REG_STATUS, &status);
+		if (err) {
+			goto unlock;
+		}
+
+		if (((status & RV3028_STATUS_AF) != 0U && alarm_callback != NULL) ||
+		    ((status & RV3028_STATUS_UF) != 0U && update_callback != NULL)) {
+			/*
+			 * Another interrupt occurred while servicing this one, process current
+			 * callback(s) and yield.
+			 */
+			k_sem_give(&data->int_sem);
+		}
+
+unlock:
+		k_mutex_unlock(&data->lock);
+
+		if (alarm_callback != NULL) {
+			alarm_callback(dev, 0U, alarm_user_data);
+			alarm_callback = NULL;
+		}
+
+		if (update_callback != NULL) {
+			update_callback(dev, update_user_data);
+			update_callback = NULL;
+		}
+	}
+}
+
+static void rv3028_isr_callback_handler(const struct device *port, struct gpio_callback *cb,
+                                        gpio_port_pins_t pins)
+{
+	struct rv3028_data *data = CONTAINER_OF(cb, struct rv3028_data, int_callback);
+
+	ARG_UNUSED(port);
+	ARG_UNUSED(pins);
+
+	k_sem_give(&data->int_sem);
+}
+
+#endif /* RV3028_INT_GPIOS_IN_USE */
+
+static int rv3028_set_time(const struct device *dev, const struct rtc_time *timeptr)
+{
+	struct rv3028_data *data = dev->data;
+	uint8_t date[7];
+	int err;
+
+	if (timeptr == NULL ||
+	    (timeptr->tm_year < RV3028_YEAR_OFFSET) ||
+	    (timeptr->tm_year > (RV3028_YEAR_OFFSET + 99))) {
```

Comment:
You may want to additionally validate the time using https://github.com/zephyrproject-rtos/zephyr/blob/3570408db556fbb53cd775ae3d2e44553e28cb31/drivers/rtc/rtc_utils.h#L26